new_dosi_adaptation.nlogo

```diff
@@ -0,0 +1,725 @@
+    create-input-good-firms num-firms[
+      set shape "truck"
+      setxy random-xcor random-ycor
+      setup-input-table input-data
+      set liquid-assets 0
+      set profits ts-create["Profits"]
+      set preferred-index one-of (table:keys INDICES)
+      set inputs-received? false
+      set done-producing? false
+      set firm-type good-produced
+      setup-input-variables capacity
+    ]
+  ]
+end
+
+to setup-input-table [input-data] ;sets up the input table for a given firm, also based on info from a csv file
+  set input-information table:make
+  foreach input-data[i ->
+    let current-input-data table:make
+    table:put current-input-data "Marginal-productivity" (item 1 i)
+    table:put current-input-data "Current-stock" 0
+    table:put current-input-data "Pending-orders" 0
+    let input-type (item 0 i)
+    let finished-negotiating false
+    let average-unit-cost 0
+    if input-type = 0[
+      set finished-negotiating true
+      set average-unit-cost WAGE-RATE
+    ]
+    table:put current-input-data "Average unit cost" average-unit-cost
+    table:put current-input-data "Agreements-negotiated?" finished-negotiating
+    table:put input-information input-type current-input-data
+  ]
+end
+
+to setup-consumer-variables [num-firms] ;sets up the consumer good firm specific variables
+  set market-share (1 / num-firms)
+  set estimated-demand 0
+  set desired-production 0
+  set inventories 0
+  set competitiveness 0
+  set price 0
+  set actual-demand 0
+  set quantity-sold 0
+end
+
+to setup-input-variables [production-capacity-info] ;sets up the input firms specific variables
+  set quantity-ordered 0
+  set production-capacity production-capacity-info
+end
+
+to-report uses-input? [good] ;run by a turtle, reports true if the turtle uses the given input
+  report table:has-key? input-information good
+end
+
+to-report agreements-negotiated? [good] ;run by a turtle, returns true if the agreements for a particular good have been negotiated
+  let good-information table:get input-information good
+  report table:get good-information "Agreements-negotiated?"
+end
+
+to-report done-negotiating? ;run by a turtle, reports true if the turtle is done negotiating all agreements
+  let total-counter 0
+  let negotiated-counter 0
+  foreach table:keys input-information [k ->
+    if agreements-negotiated? k [
+      set negotiated-counter (negotiated-counter + 1)
+    ]
+    set total-counter (total-counter + 1)
+  ]
+  ifelse total-counter = negotiated-counter[
+    report true
+  ][
+    report false
+  ]
+end
+
+to-report number-firms [f] ;run by the observer, reports the total number of firms for a given firm type
+  let information table:get FIRM-INFO f
+  report table:get information "Number"
+end
+
+to-report current-negotiating-goods [good-table] ;run by the observer, reports a list of all the goods that firms can negotiate for
+  let ready-goods (list)
+  foreach table:keys good-table[k ->
+    if table:get good-table k [
```

Comment:
I don't understand why you need this if statement since you just got all these keys from the good-table. Assuming you don't need it, then I think you can just do `let ready-goods table:keys good-table`